pkg/controller/plan/adapter/ovirt/client.go

```diff
@@ -460,89 +474,76 @@ func (r Client) removePrecopies(precopies []planapi.Precopy, vmService *ovirtsdk
-				if !found {
-					r.Log.Error(err, "disks not found")
-					err = liberr.Wrap(err, "Disks not found")
-					return false, err
-				}
-				for _, disk := range snapshotDisks.Slice() {
-					if transferred, err := r.isDiskBeingTransferred(disk.MustId()); transferred {
-						r.Log.Info("Disk is being transferred, retrying...")
-						return false, nil
-					} else if err != nil {
-						err = liberr.Wrap(err)
-						r.Log.Error(err, "Error checking if disk is being transferred")
-						return false, nil
-					} else {
-						// Try to remove the snapshot again
-						_, err = snapService.Remove().Query("correlation_id", correlationID).Send()
-						if err != nil {
-							err = liberr.Wrap(err)
-							r.Log.Error(err, "Error removing snapshot")
-							return true, err
-						}
-					}
-				}
-				return false, nil
-			})
-			if err != nil {
-				err = liberr.Wrap(err)
-				r.Log.Error(err, "Error waiting for snapshot removal")
-				return
-			}
-
-		}
-
 		for {
-			jobs, err := r.getJobs(correlationID)
-			if err != nil {
-				err = liberr.Wrap(err)
-				r.Log.Error(err, "Error waiting for snapshot removal")
-				break
-			}
-			if allJobsFinished(jobs) {
-				break
-			}
-
 			select {
 			case <-time.After(time.Duration(settings.Settings.Migration.SnapshotRemovalTimeout) * time.Minute):
 				r.Log.Info("Timeout waiting for snapshot removal")
 				return
 			default:
 				time.Sleep(time.Duration(settings.Settings.Migration.SnapshotStatusCheckRate) * time.Second)
 			}
+			// Try to remove the snapshot
+			_, err := snapService.Remove().Query("correlation_id", correlationID).Send()
+			if err != nil {
+				if strings.Contains(err.Error(), "Cannot remove Snapshot") {
+					err = web.ConflictError{
+						Provider: r.Source.Provider,
+						Err:      err,
+					}
+					r.Log.Error(err, "ConflictError failed to remove snapshot", "snapshotID", snapshotID)
+				} else {
+					err = liberr.Wrap(err)
+					r.Log.Error(err, "Remove request for snapshot failed", "snapshotID", snapshotID)
+
+					snapshotDisksResponse, err := snapService.DisksService().List().Send()
+					if err != nil {
+						r.Log.Error(err, "Request for snapshotDisksResponse failed")
+						break
+					}
+					snapshotDisks, found := snapshotDisksResponse.Disks()
+					if !found {
+						r.Log.Error(err, "Disks not found")
+						continue
+					}
+					for _, disk := range snapshotDisks.Slice() {
+						if transferred, err := r.isDiskBeingTransferred(disk.MustId()); transferred {
+							r.Log.Info("Disk is being transferred, retrying...", "disk", disk)
+						} else if err != nil {
+							err = liberr.Wrap(err)
+							r.Log.Error(err, "Error checking if disk is being transferred")
+						}
+					}
+				}
+			} else {
+				// Check the events of the snapshot
+				finished, err := r.isSnapshotRemovalFinished(correlationID)
+				if err != nil {
+					err = liberr.Wrap(err)
+					r.Log.Error(err, "Error waiting for snapshot removal")
```

Comment:
this assumes that later the problem would be resolved but since we're in a loop here (and so even if the user cancels the migration we won't stop), maybe we should rather stop in this case

Reply:
Ahh, the error log does not have the best description. It should be something like `Error occurred while gathering events about the snapshot removal`. But you are right that we should rather stop in this case

Reply:
I try to recall what I meant here - it was late yesterday when I wrote it
I think I missed the fact that it's within the loop at should also terminate when the timeout is reached..
but about the timeout, we have:
```golang
                        select {
			case <-time.After(time.Duration(settings.Settings.Migration.SnapshotRemovalTimeout) * time.Minute):
				r.Log.Info("Timeout waiting for snapshot removal")
				return
			default:
				time.Sleep(time.Duration(settings.Settings.Migration.SnapshotStatusCheckRate) * time.Second)
			}
```
wouldn't it always get to the `default` part? I mean, `case <-time.After(time.Duration(settings.Settings.Migration.SnapshotRemovalTimeout) * time.Minute):` is supposed to check whether the time since the call to remove-snapshot exceeded `time.Duration(settings.Settings.Migration.SnapshotRemovalTimeout) * time.Minute)` no?
@mnecas @bennyz ^

Reply:
indeed, I missed the noting the start time

Reply:
Updated the timeout logic please check it out